doc/dd_keys.3
'\" t
.\"     Title: dd_keys
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 12/14/2016
.\"    Manual: DoubleDecker Manual
.\"    Source: DoubleDecker 0.5.0
.\"  Language: English
.\"
.TH "DD_KEYS" "3" "12/14/2016" "DoubleDecker 0\&.5\&.0" "DoubleDecker Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
dd_keys \- Class holding client keys
.SH "SYNOPSIS"
.sp
.nf
//  This API is a draft, and may change without notice\&.
#ifdef DD_BUILD_DRAFT_API
//  *** Draft method, for development use, may change without warning ***
//  Create a keys object from a JSON keyfile file
DD_EXPORT dd_keys_t *
    dd_keys_new (const char *filename);

//  *** Draft method, for development use, may change without warning ***
//  Destroy a DoubleDecker JSON keyfile object
DD_EXPORT void
    dd_keys_destroy (dd_keys_t **self_p);

//  *** Draft method, for development use, may change without warning ***
//  Get the client keys
DD_EXPORT zhash_t *
    dd_keys_clients (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Is the key object representing a public keyset?
DD_EXPORT bool
    dd_keys_ispublic (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Get the precalculated client \- client key
DD_EXPORT const byte *
    dd_keys_custboxk (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Get the keyset customer hash
DD_EXPORT const char *
    dd_keys_hash (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Get the public key
DD_EXPORT const byte *
    dd_keys_pub (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Get the precalculated client \- broker key
DD_EXPORT const byte *
    dd_keys_ddboxk (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Get the broker public key
DD_EXPORT const byte *
    dd_keys_ddpub (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Get the precalculated  public tenant key
DD_EXPORT const byte *
    dd_keys_pubboxk (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Get the public key of the public tenant key
DD_EXPORT const byte *
    dd_keys_publicpub (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Get the private key
DD_EXPORT const byte *
    dd_keys_priv (dd_keys_t *self);

//  *** Draft method, for development use, may change without warning ***
//  Increment a nonce
DD_EXPORT void
    dd_keys_nonce_increment (byte *nonce, size_t length);

//  *** Draft method, for development use, may change without warning ***
//  Self test of this class\&.
DD_EXPORT void
    dd_keys_test (bool verbose);

#endif // DD_BUILD_DRAFT_API
.fi
.SH "DESCRIPTION"
.sp
dd_keys \-
.sp
Please add @discuss section in \&.\&./src/dd_keys\&.c\&.
.SH "EXAMPLE"
.PP
\fBFrom dd_keys_test method\fR. 
.sp
.if n \{\
.RS 4
.\}
.nf
//  Simple create/destroy test
dd_keys_t *self = dd_keys_new ("keys/public\-keys\&.json");
assert (self);
dd_keys_destroy (&self);
.fi
.if n \{\
.RE
.\}
.sp
.SH "AUTHORS"
.sp
The doubledecker manual was written by the authors in the AUTHORS file\&.
.SH "RESOURCES"
.sp
Main web site: \m[blue]\fB\%\fR\m[]
.sp
Report bugs to the email <\m[blue]\fBponsko@acreo\&.se\fR\m[]\&\s-2\u[1]\d\s+2>
.SH "COPYRIGHT"
.sp
Copyright (c) 2015 Pontus Sk\(:oldstr\(:om, Bertrand Pechenot This file is part of libdd, the DoubleDecker hierarchical messaging system DoubleDecker is free software; you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License (LGPL) version 2\&.1 as published by the Free Software Foundation\&. As a special exception, the Authors give you permission to link this library with independent modules to produce an executable, regardless of the license terms of these independent modules, and to copy and distribute the resulting executable under terms of your choice, provided that you also meet, for each linked independent module, the terms and conditions of the license of that module\&. An independent module is a module which is not derived from or based on this library\&. If you modify this library, you must extend this exception to your version of the library\&. DoubleDecker is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE\&. See the GNU Lesser General Public License for more details\&. You should have received a copy of the GNU Lesser General Public License along with this program\&. If not, see http://www\&.gnu\&.org/licenses/\&. LICENSE included with the doubledecker distribution\&.
.SH "NOTES"
.IP " 1." 4
ponsko@acreo.se
.RS 4
\%mailto:ponsko@acreo.se
.RE


src/dd_keys.c
/*  =========================================================================
    dd_keys - Class holding client keys

    Copyright (c) the Contributors as noted in the AUTHORS file.       
    This file is part of CZMQ, the high-level C binding for 0MQ:       
    http://czmq.zeromq.org.                                            
                                                                       
    This Source Code Form is subject to the terms of the Mozilla Public
    License, v. 2.0. If a copy of the MPL was not distributed with this
    file, You can obtain one at http://mozilla.org/MPL/2.0/.           
    =========================================================================
*/

/*
@header
    dd_keys - 
@discuss
@end
*/

#include "dd_classes.h"
//  Structure of our class

struct _dd_keys_t {
    // crypto_box_PUBLICKEYBYTES / crypto_box_SECRETKEYBYTES
    // 32U arrays
    // Public key of "public" tenant
    uint8_t *publicpubkey;
    // Broker public key
    uint8_t *ddpubkey;
    // Client private key
    uint8_t *privkey;
    // Client public key
    uint8_t *pubkey;

    // crypto_box_BEFORENMBYTES
    // 32U arrays
    // Broker symmetric key
    uint8_t *ddboxk;
    // Tenant symmetric key
    uint8_t *custboxk;
    // Tenant-Public symmetric key
    uint8_t *pubboxk;
    int ispublic;
    char *hash;
    zhash_t *clientkeys;
};




//  --------------------------------------------------------------------------
//  Destroy the dd_keys
void
dd_keys_destroy(dd_keys_t **self_p) {
    assert(self_p);
    if (*self_p) {
        dd_keys_t *self = *self_p;
        free(self->publicpubkey);
        free(self->ddpubkey);
        free(self->privkey);
        free(self->pubkey);
        free(self->ddboxk);
        free(self->custboxk);
        free(self->pubboxk);
        free(self->hash);
        zhash_destroy(&self->clientkeys);
        free(self);
        *self_p = NULL;
    }
}

//  --------------------------------------------------------------------------
//  Self test of this class

void
dd_keys_test (bool verbose)
{
    printf (" * dd_keys: ");

    //  @selftest
    //  Simple create/destroy test
    dd_keys_t *self = dd_keys_new ("keys/public-keys.json");
    assert (self);
    dd_keys_destroy (&self);
    //  @end
    printf ("OK\n");
}

//  --------------------------------------------------------------------------
//  Create a new dd_keys from JSON file, for a customer
//  Returns NULL on failure



dd_keys_t *
dd_keys_new(const char *filename) {
    assert(filename);
    FILE *fp;
    int retval;

    dd_keys_t *ddkeys;
    struct stat stats;

    fp = fopen(filename, "r");
    if (fp == NULL) {
        printf("Error opening file: %s, exiting..\n", filename);
        assert(fp);
    }
    retval = stat(filename, &stats);
    if (retval != 0) {
        printf("Could not stat %s, exiting..\n", filename);
        fclose(fp);
        assert(retval);
        return NULL;
    }
    char *data = (char *)calloc(1, stats.st_size + 1);
    if (data == NULL) {
        printf("Error allocating memory\n");
        fclose(fp);
        return NULL;
    }
    retval = fread(data, stats.st_size, 1, fp);
    if (retval < 1) {
        printf("Error reading file\n");
        fclose(fp);
        free(data);
        return NULL;
    }
    struct json_object *parse_result = json_tokener_parse((char *)data);
    free(data);
    fclose(fp);
    //  const char *unjson = json_object_get_string(parse_result);

    base64_decodestate state_in;
    ddkeys = (dd_keys_t *)calloc(1, sizeof(dd_keys_t));
    ddkeys->clientkeys = zhash_new();

    json_object_object_foreach(parse_result, key, val) {
        if (val == NULL); // added to avoid -Werror=pedantic        
        if (streq(key, "public")) {
            ddkeys->ispublic = true;
        }
    }

    if (ddkeys->ispublic) {
        json_object_object_foreach(parse_result, key, val) {
            if (strcmp(key, "public") == 0) {
                //enum json_type o_type = json_object_get_type(val);
                json_object_object_foreach(val, key2, val2) {
//                    enum json_type o_type2 = json_object_get_type(val2);
                    if (strcmp(key2, "publicpubkey") == 0) {
                        base64_init_decodestate(&state_in);
                        ddkeys->publicpubkey = (unsigned char *)calloc(1, 33);
                        retval = base64_decode_block(
                                json_object_get_string(val2), json_object_get_string_len(val2),
                                (char *)ddkeys->publicpubkey, &state_in);
                        if (retval != 32) {
                            fprintf(stderr, "Error during base64_decode of %s\n", key2);
                            return NULL;
                        }
                    } else if (strcmp(key2, "ddpubkey") == 0) {
                        base64_init_decodestate(&state_in);
                        ddkeys->ddpubkey = (unsigned char *)calloc(1, 33);
                        retval = base64_decode_block(json_object_get_string(val2),
                                                     json_object_get_string_len(val2),
                                                     (char *)ddkeys->ddpubkey, &state_in);
                        if (retval != 32) {
                            fprintf(stderr, "Error during base64_decode of %s\n", key2);
                            return NULL;
                        }
                    } else if (strcmp(key2, "privkey") == 0) {
                        base64_init_decodestate(&state_in);
                        ddkeys->privkey = (unsigned char *)calloc(1, 33);
                        retval = base64_decode_block(json_object_get_string(val2),
                                                     json_object_get_string_len(val2),
                                                     (char *)ddkeys->privkey, &state_in);
                        if (retval != 32) {
                            fprintf(stderr, "Error during base64_decode of %s\n", key2);
                            return NULL;
                        }
                    } else if (strcmp(key2, "pubkey") == 0) {
                        base64_init_decodestate(&state_in);
                        ddkeys->pubkey = (unsigned char *)calloc(1, 33);
                        retval = base64_decode_block(json_object_get_string(val2),
                                                     json_object_get_string_len(val2),
                                                     (char *)ddkeys->pubkey, &state_in);
                        if (retval != 32) {
                            fprintf(stderr, "Error during base64_decode of %s\n", key2);
                            return NULL;
                        }

                    } else if (strcmp(key2, "hash") == 0) {
                        ddkeys->hash = strdup(json_object_get_string(val2));
                    } else {
                    }
                }
            }
        }

        // do a second pass to find other keys
        json_object_object_foreach(parse_result, key3, val3) {
            if (strcmp(key3, "public") != 0) {
                const char *customer_name = NULL;
                const char *customer_pubkey = NULL;
                json_object_object_foreach(val3, key4, val4) {
                    // printf("Key %s : val
                    // %s\n",key4,json_object_get_string(val4));
                    if (strcmp(key4, "r") == 0) {
                        customer_name = json_object_get_string(val4);
                    } else if (strcmp(key4, "pubkey") == 0) {
                        customer_pubkey = json_object_get_string(val4);
                    }
                    unsigned char pubkey[33];
                    // printf("customer name = %s, customer_pubkey =
                    // %s\n",customer_name,customer_pubkey);
                    if (customer_name && customer_pubkey) {
                        base64_init_decodestate(&state_in);
                        retval =
                                base64_decode_block(customer_pubkey, strlen(customer_pubkey),
                                                    (char *)pubkey, &state_in);
                        unsigned char *precalck =
                                (unsigned char *)calloc(1, crypto_box_BEFORENMBYTES);
                        retval = crypto_box_beforenm(precalck, pubkey, ddkeys->privkey);
                        zhash_insert(ddkeys->clientkeys, customer_name, precalck);
                        customer_pubkey = NULL;
                        customer_name = NULL;
                    }
                }
            }
        }
    } else { // Not public!
        json_object_object_foreach(parse_result, key2, val2) {

            if (strcmp(key2, "publicpubkey") == 0) {
                base64_init_decodestate(&state_in);
                ddkeys->publicpubkey = (unsigned char *)calloc(1, 33);
                retval = base64_decode_block(json_object_get_string(val2),
                                             json_object_get_string_len(val2),
                                             (char *)ddkeys->publicpubkey, &state_in);
                if (retval != 32) {
                    fprintf(stderr, "Error during base64_decode of %s\n", key2);
                    return NULL;
                }
            } else if (strcmp(key2, "ddpubkey") == 0) {
                base64_init_decodestate(&state_in);
                ddkeys->ddpubkey = (unsigned char *)calloc(1, 33);
                retval = base64_decode_block(json_object_get_string(val2),
                                             json_object_get_string_len(val2),
                                             (char *)ddkeys->ddpubkey, &state_in);
                if (retval != 32) {
                    fprintf(stderr, "Error during base64_decode of %s\n", key2);
                    return NULL;
                }
            } else if (strcmp(key2, "privkey") == 0) {
                base64_init_decodestate(&state_in);
                ddkeys->privkey = (unsigned char *)calloc(1, 33);
                retval = base64_decode_block(json_object_get_string(val2),
                                             json_object_get_string_len(val2),
                                             (char *)ddkeys->privkey, &state_in);
                if (retval != 32) {
                    fprintf(stderr, "Error during base64_decode of %s\n", key2);
                    return NULL;
                }
            } else if (strcmp(key2, "pubkey") == 0) {
                base64_init_decodestate(&state_in);
                ddkeys->pubkey = (unsigned char *)calloc(1, 33);
                retval = base64_decode_block(json_object_get_string(val2),
                                             json_object_get_string_len(val2),
                                             (char *)ddkeys->pubkey, &state_in);
                if (retval != 32) {
                    fprintf(stderr, "Error during base64_decode of %s\n", key2);
                    return NULL;
                }

            } else if (strcmp(key2, "hash") == 0) {
                ddkeys->hash = strdup(json_object_get_string(val2));
            }
        }
    }

    // json_object_delete(parse_result);
    // TODO: how to free the memory used by the json parser here?
    json_object_put(parse_result);
    // pre-calculate the shared keys for the DD bus itself
    ddkeys->ddboxk = (unsigned char *)calloc(1, crypto_box_BEFORENMBYTES);
    retval =
            crypto_box_beforenm(ddkeys->ddboxk, ddkeys->ddpubkey, ddkeys->privkey);

    // pre-calculate the shared keys for the particular customer
    ddkeys->custboxk = (unsigned char *)calloc(1, crypto_box_BEFORENMBYTES);
    retval =
            crypto_box_beforenm(ddkeys->custboxk, ddkeys->pubkey, ddkeys->privkey);

    // pre-calculate the shared keys for the public customer
    ddkeys->pubboxk = (unsigned char *)calloc(1, crypto_box_BEFORENMBYTES);
    retval = crypto_box_beforenm(ddkeys->pubboxk, ddkeys->publicpubkey,
                                 ddkeys->privkey);

    return ddkeys;
}
zhash_t *
dd_keys_clients(dd_keys_t *self) { return self->clientkeys; }
bool
dd_keys_ispublic(dd_keys_t *self) { return self->ispublic; }
const unsigned char *
dd_keys_custboxk(dd_keys_t *self) {
    return self->custboxk;
}
const char *
dd_keys_hash(dd_keys_t *self) { return self->hash; }
const uint8_t *
dd_keys_pub(dd_keys_t *self) { return self->pubkey; }
const uint8_t *
dd_keys_ddboxk(dd_keys_t *self) { return self->ddboxk; }
const uint8_t *
dd_keys_ddpub(dd_keys_t *self) { return self->ddpubkey; }
const uint8_t *
dd_keys_pubboxk(dd_keys_t *self) { return self->pubboxk; }
const uint8_t *
dd_keys_publicpub(dd_keys_t *self) { return self->publicpubkey; }
const unsigned char *
dd_keys_priv(dd_keys_t *self) { return self->privkey; }



void dd_keys_nonce_increment(byte *n, size_t nlen) {
    size_t i = 0U;
    uint_fast16_t c = 1U;

#ifdef HAVE_AMD64_ASM
    uint64_t t64, t64_2;
  uint32_t t32;

  if (nlen == 12U) {
    __asm__ __volatile__("xorq %[t64], %[t64] \n"
                         "xorl %[t32], %[t32] \n"
                         "stc \n"
                         "adcq %[t64], (%[out]) \n"
                         "adcl %[t32], 8(%[out]) \n"
                         : [t64] "=&r"(t64), [t32] "=&r"(t32)
                         : [out] "D"(n)
                         : "memory", "flags", "cc");
    return;
  } else if (nlen == 24U) {
    __asm__ __volatile__("movq $1, %[t64] \n"
                         "xorq %[t64_2], %[t64_2] \n"
                         "addq %[t64], (%[out]) \n"
                         "adcq %[t64_2], 8(%[out]) \n"
                         "adcq %[t64_2], 16(%[out]) \n"
                         : [t64] "=&r"(t64), [t64_2] "=&r"(t64_2)
                         : [out] "D"(n)
                         : "memory", "flags", "cc");
    return;
  } else if (nlen == 8U) {
    __asm__ __volatile__("incq (%[out]) \n"
                         :
                         : [out] "D"(n)
                         : "memory", "flags", "cc");
    return;
  }
#endif
    for (; i < nlen; i++) {
        c += (uint_fast16_t)n[i];
        n[i] = (unsigned char)c;
        c >>= 8;
    }
}

/* TODO, API for key generation


  generate broker keys
   tenants = null
   privkey = something
   pubkey = something
   tenantkeys = null
   hash = something
   cookie = something
   dd_broker_keys_t *broker_keys = dd_broker_keys_new();


   generate public tenant and add to broker_keys
   return the public tenant keys

   dd_keys_t* public_key = dd_keys_add_public(
    broker_keys);

   generate private tenant and add to broker keys
   return the private tenant keys

   dd_keys_t* private_key = dd_keys_add_private(
   broker_keys, public_key, "tenant");

   char * dd_keys_json(public_key);
   fprintf(.. dd_keys_json());
   fprintf(.. dd_keys_json());
   fprintf(.. dd_broker_keys_json());
*/
